prusti-viper/src/encoder/procedure_encoder.rs

```diff
@@ -2444,12 +2511,202 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
+            &destination.as_ref().unwrap().0,
+            ArrayAccessKind::Mutable(None, location),
+        )?;
+        stmts.extend(encode_stmts);
+        if !lhs_ty.is_slice() { return Err(EncodingError::unsupported(
+            format!("Non-slice LHS type '{:?}' not supported yet", lhs_ty)
+        ))}
+
+        stmts.extend(self.encode_havoc(&encoded_lhs));
+        stmts.push(vir_stmt!{ inhale [vir::Expr::pred_permission(encoded_lhs.clone(), vir::PermAmount::Read).unwrap()] });
+
+        let lhs_slice_ty = lhs_ty.peel_refs();
+        let lhs_slice_expr = self.encoder.encode_value_expr(encoded_lhs.clone(), lhs_ty)?;
+
+        let base_seq = self.mir_encoder.encode_operand_place(&args[0])?.unwrap();
+        let base_seq_ty = self.mir_encoder.get_operand_ty(&args[0]);
+
+        // base_seq is expected to be ref$Array$.. or ref$Slice$.., but lookup_pure wants the
+        // contained Array$../Slice$..
+        let base_seq_expr = self.encoder.encode_value_expr(base_seq, base_seq_ty)?;
+
+        let j = vir_local!{ j: Int };
+        let tymap = HashMap::new();
+        let rhs_lookup_j = match base_seq_ty {
+            a if a.peel_refs().is_array() => {
+                let enc_array_types = self.encoder.encode_array_types(a.peel_refs())?;
+                let elem_snap_ty = self.encoder.encode_snapshot_type(enc_array_types.elem_ty_rs, &tymap)?;
+                enc_array_types.encode_lookup_pure_call(
+                    self.encoder,
+                    base_seq_expr.clone(),
+                    j.clone().into(),
+                    elem_snap_ty,
+                )
+            }
+            s if s.is_slice() => {
+                let enc_slice_types = self.encoder.encode_slice_types(s.peel_refs())?;
+                let elem_snap_ty = self.encoder.encode_snapshot_type(enc_slice_types.elem_ty_rs, &tymap)?;
+                enc_slice_types.encode_lookup_pure_call(
+                    self.encoder,
+                    base_seq_expr.clone(),
+                    j.clone().into(),
+                    elem_snap_ty,
+                )
+            }
+            _ => return Err(EncodingError::unsupported(
+                format!("Slicing is only supported for arrays/slices currently, not '{:?}'", base_seq_ty)
+            ))
+        };
+
+        let encoded_idx = self.mir_encoder.encode_operand_place(&args[1])?.unwrap();
+        trace!("idx: {:?}", encoded_idx);
+        let idx_ty = self.mir_encoder.get_operand_ty(&args[1]);
+        let idx_ident = self.encoder.env().tcx().def_path_str(idx_ty.ty_adt_def().unwrap().did);
+        trace!("ident: {}", idx_ident);
+
+        self.slice_created_at.insert(location, encoded_lhs);
+
+        // TODO: what do we actually do here? this seems a littly hacky.
+        // there's fields like _5.f$start.val_int on `encoded_idx`, it just feels hacky to
+        // manually re-do them here when we probably just encoded the type and the
+        // construction of the fields..
```

Comment:
```suggestion
        // TODO: there's fields like _5.f$start.val_int on `encoded_idx`, it just feels hacky to
        // manually re-do and hardcode them here when we probably just encoded the type
        // and the construction of the fields.
```

Reply:
So yes, the relying on "sdtlib internals" is the hacky thing imo. Though if these are unlikely to change then maybe it's ok to hardcode, and just remove this comment?

Reply:
Yeah I guess for now this is ok. I wonder if eventually we can "decouple" from this somehow, but slice indexing seems to be very deeply embedded in the stdlib. There is the [`Index`](https://doc.rust-lang.org/std/ops/trait.Index.html#impl-Index%3CI%3E-1) implementation for slices, which itself requires a [`SliceIndex`](https://doc.rust-lang.org/std/slice/trait.SliceIndex.html) implementation. The concrete `Range*` types are mentioned [there](https://doc.rust-lang.org/std/slice/trait.SliceIndex.html#impl-SliceIndex%3Cstr%3E). Maybe a long time down the line we could extern spec these trait implementations in `prusti-stdlib` somehow, instead of hard-coding the `Range*` types here.